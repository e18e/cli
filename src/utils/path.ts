export function normalizePath(path: string): string {
  return path.replaceAll('\\', '/');
}
