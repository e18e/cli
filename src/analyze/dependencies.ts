import colors from 'picocolors';
import {analyzePackageModuleType} from '../compute-type.js';
import type {
  PackageJsonLike,
  ReportPluginResult,
  Message,
  Stats
} from '../types.js';
import {FileSystem} from '../file-system.js';
import {normalizePath} from '../utils/path.js';

interface DependencyNode {
  name: string;
  version: string;
  // TODO (43081j): make this an array or something structured one day
  path: string; // Path in dependency tree (e.g., "root > package-a > package-b")
  parent?: string; // Parent package name
  depth: number; // Depth in dependency tree
  packagePath: string; // File system path to package.json
}

interface DuplicateDependency {
  name: string;
  versions: DependencyNode[];
  severity: 'exact' | 'conflict' | 'resolvable';
  potentialSavings?: number;
  suggestions?: string[];
}

/**
 * Detects duplicate dependencies from a list of dependency nodes
 */
function detectDuplicates(
  dependencyNodes: DependencyNode[]
): DuplicateDependency[] {
  const duplicates: DuplicateDependency[] = [];
  const packageGroups = new Map<string, DependencyNode[]>();

  // Group dependencies by name
  for (const node of dependencyNodes) {
    if (!packageGroups.has(node.name)) {
      packageGroups.set(node.name, []);
    }
    packageGroups.get(node.name)?.push(node);
  }

  // Find packages with multiple versions
  for (const [packageName, nodes] of packageGroups) {
    if (nodes.length > 1) {
      const duplicate = analyzeDuplicate(packageName, nodes);
      if (duplicate) {
        duplicates.push(duplicate);
      }
    }
  }

  return duplicates;
}

/**
 * Analyzes a group of nodes for the same package to determine duplicate type
 */
function analyzeDuplicate(
  packageName: string,
  nodes: DependencyNode[]
): DuplicateDependency | null {
  // Skip root package
  if (packageName === 'root' || nodes.some((n) => n.name === 'root')) {
    return null;
  }

  const uniqueVersions = new Set(nodes.map((n) => n.version));

  let severity: 'exact' | 'conflict' | 'resolvable';

  // If more than one version, it's a conflict
  if (uniqueVersions.size === 1) {
    severity = 'exact';
  } else {
    severity = 'conflict';
  }

  return {
    name: packageName,
    versions: nodes,
    severity,
    potentialSavings: calculatePotentialSavings(nodes),
    suggestions: generateSuggestions(nodes)
  };
}

/**
 * Calculates potential savings from deduplication
 */
function calculatePotentialSavings(nodes: DependencyNode[]): number {
  // For now, return a simple estimate based on number of duplicates
  // TODO: Implement actual size calculation
  return nodes.length - 1;
}

/**
 * Generates suggestions for resolving duplicates
 */
function generateSuggestions(nodes: DependencyNode[]): string[] {
  const suggestions: string[] = [];

  // Group by version to identify the most common version
  const versionCounts = new Map<string, number>();
  for (const node of nodes) {
    versionCounts.set(node.version, (versionCounts.get(node.version) || 0) + 1);
  }

  const mostCommonVersion = Array.from(versionCounts.entries()).sort(
    (a, b) => b[1] - a[1]
  )[0];

  if (mostCommonVersion && mostCommonVersion[1] > 1) {
    suggestions.push(
      `Consider standardizing on version ${mostCommonVersion[0]} (used by ${mostCommonVersion[1]} dependencies)`
    );
  }

  // Suggest checking for newer versions of consuming packages
  const uniqueParents = new Set(nodes.map((n) => n.parent).filter(Boolean));
  if (uniqueParents.size > 1) {
    suggestions.push(
      `Check if newer versions of consuming packages (${Array.from(uniqueParents).join(', ')}) would resolve this duplicate`
    );
  }

  return suggestions;
}

/**
 * Attempts to parse a `package.json` file
 */
async function parsePackageJson(
  fileSystem: FileSystem,
  path: string
): Promise<PackageJsonLike | null> {
  try {
    return JSON.parse(await fileSystem.readFile(path));
  } catch {
    return null;
  }
}

// Keep the existing tarball analysis for backward compatibility
export async function runDependencyAnalysis(
  fileSystem: FileSystem
): Promise<ReportPluginResult> {
  const packageFiles = await fileSystem.listPackageFiles();
  const rootDir = await fileSystem.getRootDir();
  const messages: Message[] = [];

  // Find root package.json
  const pkg = await parsePackageJson(fileSystem, '/package.json');

  if (!pkg) {
    throw new Error('No package.json found.');
  }

  const installSize = await fileSystem.getInstallSize();
  const prodDependencies = Object.keys(pkg.dependencies || {}).length;
  const devDependencies = Object.keys(pkg.devDependencies || {}).length;
  const stats: Stats = {
    name: pkg.name,
    version: pkg.version,
    installSize,
    dependencyCount: {
      production: prodDependencies,
      development: devDependencies,
      esm: 0,
      cjs: 0,
      duplicate: 0
    }
  };

  let cjsDependencies = 0;
  let esmDependencies = 0;
  const dependencyNodes: DependencyNode[] = [];

  // Recursively traverse dependencies
  async function traverse(
    packagePath: string,
    parent: string | undefined,
    depth: number,
    pathInTree: string
  ) {
    const depPkg = await parsePackageJson(fileSystem, packagePath);
    if (!depPkg || !depPkg.name) return;

    // Record this node
    dependencyNodes.push({
      name: depPkg.name,
      version: depPkg.version || 'unknown',
      path: pathInTree,
      parent,
      depth,
      packagePath
    });

    // Only count CJS/ESM for non-root packages
    if (depth > 0) {
      const type = analyzePackageModuleType(depPkg);
      if (type === 'cjs') cjsDependencies++;
      if (type === 'esm') esmDependencies++;
      if (type === 'dual') {
        cjsDependencies++;
        esmDependencies++;
      }
    }

    // Traverse dependencies
    const allDeps = {...depPkg.dependencies, ...depPkg.devDependencies};
    for (const depName of Object.keys(allDeps)) {
      let packageMatch = packageFiles.find((packageFile) =>
        normalizePath(packageFile).endsWith(`/node_modules/${depName}/package.json`)
      );

      if (!packageMatch) {
        for (const packageFile of packageFiles) {
          const depPkg = await parsePackageJson(fileSystem, packageFile);
          if (depPkg !== null && depPkg.name === depName) {
            packageMatch = packageFile;
            break;
          }
        }
      }

      if (packageMatch) {
        await traverse(
          packageMatch,
          depPkg.name,
          depth + 1,
          pathInTree + ' > ' + depName
        );
      }
    }
  }

  // Start traversal from root
  await traverse('/package.json', undefined, 0, 'root');

  // Collect all dependency instances for duplicate detection
  // This ensures we find all versions, even those in nested node_modules
  // TODO (43081j): don't do this. we're re-traversing most files just to
  // find the ones that don't exist in the parent package's dependency list.
  // there must be a better way
  for (const file of packageFiles) {
    const rootPackageJsonPath = normalizePath(rootDir) + '/package.json';
    if (file === rootPackageJsonPath) {
      continue;
    }

    try {
      const depPkg = await parsePackageJson(fileSystem, file);
      if (!depPkg || !depPkg.name) {
        continue;
      }

      // Check if we already have this exact package in our dependency nodes
      const alreadyExists = dependencyNodes.some(
        (node) => node.packagePath === file
      );

      if (!alreadyExists) {
        // Extract path information from the file path
        const normalizedFile = normalizePath(file);
        const pathParts = normalizedFile.split('/node_modules/');
        if (pathParts.length > 1) {
          const packageDirName = pathParts[pathParts.length - 1].replace(
            '/package.json',
            ''
          );
          const parentDirName = pathParts[pathParts.length - 2]
            ?.split('/')
            .pop();

          dependencyNodes.push({
            name: depPkg.name,
            version: depPkg.version || 'unknown',
            path: packageDirName,
            parent: parentDirName,
            depth: pathParts.length - 1,
            packagePath: file
          });
        }
      }
    } catch {
      // Skip invalid package.json files
    }
  }

  // Detect duplicates from the collected dependency nodes
  const duplicateDependencies = detectDuplicates(dependencyNodes);

  stats.dependencyCount.cjs = cjsDependencies;
  stats.dependencyCount.esm = esmDependencies;

  if (duplicateDependencies.length > 0) {
    stats.dependencyCount.duplicate = duplicateDependencies.length;

    for (const duplicate of duplicateDependencies) {
      const severityColor =
        duplicate.severity === 'exact' ? colors.blue : colors.yellow;

      let message = `${severityColor('[duplicate dependency]')} ${colors.bold(duplicate.name)} has ${duplicate.versions.length} installed versions:`;

      for (const version of duplicate.versions) {
        message += `\n ${colors.gray(version.version)} via ${colors.gray(version.path)}`;
      }

      if (duplicate.suggestions && duplicate.suggestions.length > 0) {
        message += '\nSuggestions:';
        for (const suggestion of duplicate.suggestions) {
          message += `    ${colors.blue('ðŸ’¡')} ${colors.gray(suggestion)}`;
        }
      }

      messages.push({
        message,
        severity: 'warning',
        score: 0
      });
    }
  }

  return {stats, messages};
}
