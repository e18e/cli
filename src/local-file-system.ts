import type {FileSystem} from './file-system.js';
import * as path from 'node:path';
import * as fs from 'node:fs/promises';
import {type Logger, pino} from 'pino';
import {fdir} from 'fdir';
import {readFile, stat} from 'node:fs/promises';

export class LocalFileSystem implements FileSystem {
  #root: string;
  #logger: Logger;

  constructor(root: string) {
    this.#root = root;
    this.#logger = pino({
      transport: {
        target: 'pino-pretty',
        options: {
          colorize: true,
          translateTime: 'SYS:standard',
          ignore: 'pid,hostname'
        }
      }
    });
  }

  async getRootDir(): Promise<string> {
    return this.#root;
  }

  async listPackageFiles(): Promise<string[]> {
    const nodeModulesPath = path.join(this.#root, 'node_modules');

    try {
      await fs.access(nodeModulesPath);
      const crawler = new fdir()
        .withFullPaths()
        .withSymlinks()
        .filter((filePath) => filePath.endsWith('/package.json'))
        .crawl(nodeModulesPath);
      return await crawler.withPromise();
    } catch {
      return [];
    }
  }

  async readFile(filePath: string): Promise<string> {
    return await readFile(path.join(this.#root, filePath), 'utf8');
  }

  async getInstallSize(): Promise<number> {
    const nodeModulesPath = path.join(this.#root, 'node_modules');

    let installSize = 0;

    try {
      await fs.access(nodeModulesPath);

      // TODO (43081j): we traverse the file system twice. Once here,
      // and once when finding the package.json files. It may be worth
      // caching some things one day so we only traverse once.
      const crawler = new fdir()
        .withFullPaths()
        .withSymlinks()
        .crawl(nodeModulesPath);
      const files = await crawler.withPromise();
      for (const filePath of files) {
        try {
          const stats = await stat(filePath);
          installSize += stats.size;
        } catch {
          this.#logger.debug('Error getting file stats for:', filePath);
        }
      }
    } catch {
      this.#logger.debug('No node_modules directory found');
    }

    return installSize;
  }
}
