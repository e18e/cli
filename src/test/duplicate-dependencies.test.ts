import {describe, it, expect, beforeEach, afterEach} from 'vitest';
import {runDependencyAnalysis} from '../analyze/dependencies.js';
import {LocalFileSystem} from '../local-file-system.js';
import {
  createTempDir,
  cleanupTempDir,
  createTestPackage,
  createTestPackageWithDependencies,
  type TestPackage
} from './utils.js';
import fs from 'node:fs/promises';
import path from 'node:path';

describe('Duplicate Dependency Detection', () => {
  let tempDir: string;
  let fileSystem: LocalFileSystem;

  beforeEach(async () => {
    tempDir = await createTempDir();
    fileSystem = new LocalFileSystem(tempDir);
  });

  afterEach(async () => {
    await cleanupTempDir(tempDir);
  });

  it('should detect exact duplicate dependencies', async () => {
    const rootPackage: TestPackage = {
      name: 'test-package',
      version: '1.0.0',
      dependencies: {
        'package-a': '1.0.0',
        'package-b': '1.0.0'
      }
    };

    const dependencies: TestPackage[] = [
      {
        name: 'package-a',
        version: '1.0.0',
        dependencies: {
          'shared-lib': '2.0.0'
        }
      },
      {
        name: 'package-b',
        version: '1.0.0',
        dependencies: {
          'shared-lib': '2.0.0'
        }
      },
      {
        name: 'shared-lib',
        version: '2.0.0'
      }
    ];

    await createTestPackageWithDependencies(tempDir, rootPackage, dependencies);

    const stats = await runDependencyAnalysis(fileSystem);

    expect(stats).toMatchSnapshot();
  });

  it('should detect version conflicts', async () => {
    const rootPackage: TestPackage = {
      name: 'test-package',
      version: '1.0.0',
      dependencies: {
        'package-a': '1.0.0',
        'package-b': '1.0.0'
      }
    };

    // Create root package
    await createTestPackage(tempDir, rootPackage, {createNodeModules: true});

    // Create package-a with shared-lib v1.0.0
    const packageADir = path.join(tempDir, 'node_modules', 'package-a');
    await fs.mkdir(packageADir);
    await createTestPackage(packageADir, {
      name: 'package-a',
      version: '1.0.0',
      dependencies: {
        'shared-lib': '1.0.0'
      }
    });

    // Create package-b with shared-lib v2.0.0
    const packageBDir = path.join(tempDir, 'node_modules', 'package-b');
    await fs.mkdir(packageBDir);
    await createTestPackage(packageBDir, {
      name: 'package-b',
      version: '1.0.0',
      dependencies: {
        'shared-lib': '2.0.0'
      }
    });

    // Create shared-lib v1.0.0
    const sharedLibV1Dir = path.join(tempDir, 'node_modules', 'shared-lib');
    await fs.mkdir(sharedLibV1Dir);
    await createTestPackage(sharedLibV1Dir, {
      name: 'shared-lib',
      version: '1.0.0'
    });

    // Create shared-lib v2.0.0 in a nested location
    const sharedLibV2Dir = path.join(
      tempDir,
      'node_modules',
      'package-b',
      'node_modules',
      'shared-lib'
    );
    await fs.mkdir(sharedLibV2Dir, {recursive: true});
    await createTestPackage(sharedLibV2Dir, {
      name: 'shared-lib',
      version: '2.0.0'
    });

    const stats = await runDependencyAnalysis(fileSystem);

    expect(stats).toMatchSnapshot();
  });

  it('should not detect duplicates when there are none', async () => {
    const rootPackage: TestPackage = {
      name: 'test-package',
      version: '1.0.0',
      dependencies: {
        'package-a': '1.0.0'
      }
    };

    const dependencies: TestPackage[] = [
      {
        name: 'package-a',
        version: '1.0.0'
      }
    ];

    await createTestPackageWithDependencies(tempDir, rootPackage, dependencies);

    const stats = await runDependencyAnalysis(fileSystem);

    expect(stats).toMatchSnapshot();
  });

  it('should generate suggestions for duplicates', async () => {
    const rootPackage: TestPackage = {
      name: 'test-package',
      version: '1.0.0',
      dependencies: {
        'package-a': '1.0.0',
        'package-b': '1.0.0'
      }
    };

    const dependencies: TestPackage[] = [
      {
        name: 'package-a',
        version: '1.0.0',
        dependencies: {
          'shared-lib': '2.0.0'
        }
      },
      {
        name: 'package-b',
        version: '1.0.0',
        dependencies: {
          'shared-lib': '2.0.0'
        }
      },
      {
        name: 'shared-lib',
        version: '2.0.0'
      }
    ];

    await createTestPackageWithDependencies(tempDir, rootPackage, dependencies);

    const stats = await runDependencyAnalysis(fileSystem);

    expect(stats).toMatchSnapshot();
  });
});
