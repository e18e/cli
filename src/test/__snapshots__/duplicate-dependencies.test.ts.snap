// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Duplicate Dependency Detection > should detect exact duplicate dependencies 1`] = `
{
  "messages": [
    {
      "message": "[34m[duplicate dependency][39m [1mshared-lib[22m has 2 installed versions:
 [90m2.0.0[39m via [90mroot > package-a > shared-lib[39m
 [90m2.0.0[39m via [90mroot > package-b > shared-lib[39m
Suggestions:    [34mðŸ’¡[39m [90mConsider standardizing on version 2.0.0 (used by 2 dependencies)[39m    [34mðŸ’¡[39m [90mCheck if newer versions of consuming packages (package-a, package-b) would resolve this duplicate[39m",
      "score": 0,
      "severity": "warning",
    },
  ],
  "stats": {
    "dependencyCount": {
      "cjs": 4,
      "development": 0,
      "duplicate": 1,
      "esm": 0,
      "production": 2,
    },
    "installSize": 244,
    "name": "test-package",
    "version": "1.0.0",
  },
}
`;

exports[`Duplicate Dependency Detection > should detect version conflicts 1`] = `
{
  "messages": [
    {
      "message": "[33m[duplicate dependency][39m [1mshared-lib[22m has 3 installed versions:
 [90m1.0.0[39m via [90mroot > package-a > shared-lib[39m
 [90m1.0.0[39m via [90mroot > package-b > shared-lib[39m
 [90m2.0.0[39m via [90mshared-lib[39m
Suggestions:    [34mðŸ’¡[39m [90mConsider standardizing on version 1.0.0 (used by 2 dependencies)[39m    [34mðŸ’¡[39m [90mCheck if newer versions of consuming packages (package-a, package-b) would resolve this duplicate[39m",
      "score": 0,
      "severity": "warning",
    },
  ],
  "stats": {
    "dependencyCount": {
      "cjs": 4,
      "development": 0,
      "duplicate": 1,
      "esm": 0,
      "production": 2,
    },
    "installSize": 292,
    "name": "test-package",
    "version": "1.0.0",
  },
}
`;

exports[`Duplicate Dependency Detection > should generate suggestions for duplicates 1`] = `
{
  "messages": [
    {
      "message": "[34m[duplicate dependency][39m [1mshared-lib[22m has 2 installed versions:
 [90m2.0.0[39m via [90mroot > package-a > shared-lib[39m
 [90m2.0.0[39m via [90mroot > package-b > shared-lib[39m
Suggestions:    [34mðŸ’¡[39m [90mConsider standardizing on version 2.0.0 (used by 2 dependencies)[39m    [34mðŸ’¡[39m [90mCheck if newer versions of consuming packages (package-a, package-b) would resolve this duplicate[39m",
      "score": 0,
      "severity": "warning",
    },
  ],
  "stats": {
    "dependencyCount": {
      "cjs": 4,
      "development": 0,
      "duplicate": 1,
      "esm": 0,
      "production": 2,
    },
    "installSize": 244,
    "name": "test-package",
    "version": "1.0.0",
  },
}
`;

exports[`Duplicate Dependency Detection > should not detect duplicates when there are none 1`] = `
{
  "messages": [],
  "stats": {
    "dependencyCount": {
      "cjs": 1,
      "development": 0,
      "duplicate": 0,
      "esm": 0,
      "production": 1,
    },
    "installSize": 47,
    "name": "test-package",
    "version": "1.0.0",
  },
}
`;
